protocol/shannon/protocol.go

```diff
@@ -0,0 +1,176 @@
+package shannon
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/pokt-network/poktroll/pkg/polylog"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	sdk "github.com/pokt-network/shannon-sdk"
+
+	"github.com/buildwithgrove/path/gateway"
+	"github.com/buildwithgrove/path/protocol"
+)
+
+// gateway package's Protocol interface is fulfilled by the Protocol struct
+// below using methods that are specific to Shannon.
+var _ gateway.Protocol = &Protocol{}
+
+// FullNode defines the set of capabilities the Shannon protocol integration needs
+// from a fullnode for sending relays.
+type FullNode interface {
+	GetServiceApps(protocol.ServiceID) ([]apptypes.Application, error)
+	// Note: Shannon returns the latest session for a service+app combination if no blockHeight is provided.
+	// This is used here because the gateway only needs the current session for any service+app combination.
+	GetSession(serviceID protocol.ServiceID, appAddr string) (sessiontypes.Session, error)
+
+	// ValidateRelayResponse validates the raw bytes returned from an endpoint (in response to a relay request) and returns the parsed response.
+	ValidateRelayResponse(supplierAddr sdk.SupplierAddress, responseBz []byte) (*servicetypes.RelayResponse, error)
+
+	// IsHealthy returns true if the FullNode instance is healthy.
+	// A LazyFullNode will always return true.
+	// A CachingFullNode will return true if it has data in app and session caches.
+	IsHealthy() bool
+
+	// GetAccountClient returns the account client from the fullnode, to be used in building relay request signers.
+	GetAccountClient() *sdk.AccountClient
+}
+
+// NewProtocol instantiates an instance of the Shannon protocol integration.
+func NewProtocol(
+	fullNode FullNode,
+	logger polylog.Logger,
+	config GatewayConfig,
+) (*Protocol, error) {
+	ownedAppsAddr, err := getCentralizedModeOwnedAppsAddr(config.OwnedAppsPrivateKeysHex)
+	if err != nil {
+		return nil, fmt.Errorf("NewProtocol: error parsing the supplied private keys: %w", err)
+	}
+
+	ownedAppsAddrIdx := make(map[string]struct{})
+	for _, appAddr := range ownedAppsAddr {
+		ownedAppsAddrIdx[appAddr] = struct{}{}
+	}
+
+	return &Protocol{
+		FullNode: fullNode,
+		Logger:   logger,
+
+		// TODO_MVP(@adshmh): verify the gateway address and private key are valid.
+		gatewayAddr:          config.GatewayAddress,
+		gatewayPrivateKeyHex: config.GatewayPrivateKeyHex,
+		gatewayMode:          config.GatewayMode,
+		ownedAppsAddr:        ownedAppsAddrIdx,
+	}, nil
+}
+
+// Protocol provides the functionality needed by the gateway package for sending a relay to a specific endpoint.
```

Comment:
Confirming that this needs to be publically exported?

Reply:
Yes, the intention here is to keep everything public, unless there is a specific reason, e.g. internal fields that should only be modified through specific business logic. All the private fields in `Protocol` could later be exposed either directly or through builder functions.